src/backend/distributed/commands/multi_copy.c

```diff
@@ -1015,14 +1015,67 @@ CanUseBinaryCopyFormatForType(Oid typeId)
 		return false;
 	}
 
-	if (typeId >= FirstNormalObjectId)
+	/*
+	 * A row type could contain any types, possibly types that don't have
+	 * the binary input and output functions defined.
+	 *
+	 * TODO: Inspect the types inside the record and check if all of them
+	 * can be binary encoded. If so, it's safe to use binary encoding.
+	 */
+	if (type_is_rowtype(typeId))
 	{
-		char typeCategory = '\0';
-		bool typePreferred = false;
+		return false;
+	}
 
-		get_type_category_preferred(typeId, &typeCategory, &typePreferred);
-		if (typeCategory == TYPCATEGORY_ARRAY ||
-			typeCategory == TYPCATEGORY_COMPOSITE)
+	HeapTuple typeTup = typeidType(typeId);
+	Form_pg_type type = (Form_pg_type) GETSTRUCT(typeTup);
+	Oid elementType = type->typelem;
+#if PG_VERSION_NUM < PG_VERSION_14
+	char typeCategory = type->typcategory;
+#endif
+	ReleaseSysCache(typeTup);
+
+	/*
+	 * Any type that is a wrapper around an element type (e.g. arrays and
+	 * ranges) require the element type to also has support for binary
+	 * encoding.
+	 */
+	if (elementType != InvalidOid)
+	{
+		if (!CanUseBinaryCopyFormatForType(elementType))
+		{
+			return false;
+		}
+	}
+
+#if PG_VERSION_NUM < PG_VERSION_14
+
+	/*
+	 * In PG versions before PG14 both array_recv and record_recv would error
+	 * out more than necessary.
+	 *
+	 * Both error out when the element types their oids don't match with the
+	 * oid in the received data. This happens pretty much always for non built
+	 * in types, because their oids differ between postgres intallations.
+	 */
+	if (typeCategory == TYPCATEGORY_ARRAY && elementType >= FirstNormalObjectId)
+	{
+		return false;
+	}
+
+	if (typeCategory == TYPCATEGORY_COMPOSITE && typeId >= FirstNormalObjectId)
```

Comment:
At the moment this check should never be reached, since we already return false for all rowtypes earlier in the function. This will hopefully be changed in the future though, so this check is kept here to make sure we don't forget to check for that later.

Reply:
Although, we have tests for this case, so maybe we can remove it and add it back when needed. @marcocitus what do you think?

Reply:
I guess it encapsulates some useful findings that would take time to rediscover, but could use a comment (or even be a comment).

Reply:
I realized the check was not working in the case of a record type containing custom types. So I removed this check and wrote a comment about it next to the TODO, to explain what a correct fix would look like.